File: WinForms/VB/SchedulerOutlookDemo/Modules/Common/ZoomBehaviour.vb
Imports System
Imports DevExpress.XtraBars
Imports DevExpress.XtraEditors.Repository

Namespace DevExpress.XtraScheduler.Demos

    Public Module ZoomBehaviour

        Private Property TrackBarItem As BarEditItem

        Private Property Scheduler As SchedulerControl

        Private ReadOnly Property TrackBarRepositoryItem As RepositoryItemZoomTrackBar
            Get
                Return If(TrackBarItem IsNot Nothing, TryCast(TrackBarItem.Edit, RepositoryItemZoomTrackBar), Nothing)
            End Get
        End Property

        Public Sub RegisterZoomTracker(ByVal trackBarItem As BarEditItem)
            ZoomBehaviour.TrackBarItem = trackBarItem
            AddHandler ZoomBehaviour.TrackBarItem.EditValueChanged, AddressOf OnTrackBarItemEditValueChanged
        End Sub

        Public Sub Attach(ByVal scheduler As SchedulerControl)
            ZoomBehaviour.Scheduler = scheduler
            AddHandler scheduler.ActiveViewChanged, AddressOf OnSchedulerActiveViewChanged
            AddHandler scheduler.DayView.PropertyChanged, AddressOf OvViewPropertyChanged
            AddHandler scheduler.WorkWeekView.PropertyChanged, AddressOf OvViewPropertyChanged
            AddHandler scheduler.FullWeekView.PropertyChanged, AddressOf OvViewPropertyChanged
            Call UpdateTrackBar()
        End Sub

        Public Sub Detach(ByVal scheduler As SchedulerControl)
            RemoveHandler scheduler.ActiveViewChanged, AddressOf OnSchedulerActiveViewChanged
            RemoveHandler scheduler.DayView.PropertyChanged, AddressOf OvViewPropertyChanged
            RemoveHandler scheduler.WorkWeekView.PropertyChanged, AddressOf OvViewPropertyChanged
            RemoveHandler scheduler.FullWeekView.PropertyChanged, AddressOf OvViewPropertyChanged
            ZoomBehaviour.Scheduler = Nothing
            Call UpdateTrackBar()
        End Sub

        Private Sub OvViewPropertyChanged(ByVal sender As Object, ByVal e As System.ComponentModel.PropertyChangedEventArgs)
            If Equals(e.PropertyName, "TimeScale") Then Call UpdateTrackBar()
        End Sub

        Private Sub OnTrackBarItemEditValueChanged(ByVal sender As Object, ByVal e As EventArgs)
            Call UpdateControl()
        End Sub

        Private Sub OnSchedulerActiveViewChanged(ByVal sender As Object, ByVal e As EventArgs)
            Call UpdateTrackBar()
        End Sub

        Private Sub UpdateControl()
            If Scheduler Is Nothing Then Return
            If TrackBarItem.EditValue Is Nothing Then Return
            Dim dayView As DayView = TryCast(Scheduler.ActiveView, DayView)
            If dayView IsNot Nothing Then dayView.TimeScale = dayView.TimeSlots(CInt(TrackBarItem.EditValue)).Value
            Dim timelineView As TimelineView = TryCast(Scheduler.ActiveView, TimelineView)
            If timelineView IsNot Nothing Then timelineView.GetBaseTimeScale().Width = CInt(TrackBarItem.EditValue)
        End Sub

        Private Sub UpdateTrackBar()
            If TrackBarItem Is Nothing OrElse TrackBarRepositoryItem Is Nothing Then Return
            If Scheduler Is Nothing Then
                TrackBarItem.Enabled = False
                TrackBarRepositoryItem.Maximum = 0
                TrackBarItem.EditValue = Nothing
                Return
            End If

            Dim dayView As DayView = TryCast(Scheduler.ActiveView, DayView)
            If dayView IsNot Nothing Then
                TrackBarRepositoryItem.Minimum = 0
                TrackBarRepositoryItem.Middle = 1
                TrackBarRepositoryItem.Maximum = dayView.TimeSlots.Count - 1
                TrackBarItem.EditValue = GetTimeSlotIndex(dayView.TimeSlots, dayView.TimeScale)
            End If

            Dim timelineView As TimelineView = TryCast(Scheduler.ActiveView, TimelineView)
            If timelineView IsNot Nothing Then
                TrackBarRepositoryItem.Maximum = 100
                TrackBarRepositoryItem.Minimum = 40
                TrackBarItem.EditValue = timelineView.GetBaseTimeScale().Width
            End If

            TrackBarItem.Enabled = timelineView IsNot Nothing OrElse dayView IsNot Nothing
        End Sub

        Private Function GetTimeSlotIndex(ByVal timeSlots As TimeSlotCollection, ByVal scale As TimeSpan) As Integer
            For i As Integer = 0 To timeSlots.Count - 1
                If timeSlots(i).Value = scale Then Return i
            Next

            Return -1
        End Function
    End Module
End Namespace

File: WinForms/CS/SchedulerOutlookDemo/Modules/Common/ZoomBehaviour.cs
using System;
using System.Linq;
using DevExpress.XtraBars;
using DevExpress.XtraEditors.Repository;

namespace DevExpress.XtraScheduler.Demos {
    public static class ZoomBehaviour {
        static BarEditItem TrackBarItem { get; set; }
        static SchedulerControl Scheduler { get; set; }
        static RepositoryItemZoomTrackBar TrackBarRepositoryItem { get { return (TrackBarItem != null) ? TrackBarItem.Edit as RepositoryItemZoomTrackBar : null; } }

        public static void RegisterZoomTracker(BarEditItem trackBarItem) {
            TrackBarItem = trackBarItem;
            TrackBarItem.EditValueChanged += OnTrackBarItemEditValueChanged;
        }

        public static void Attach(SchedulerControl scheduler) {
            Scheduler = scheduler;
            scheduler.ActiveViewChanged += OnSchedulerActiveViewChanged;
            scheduler.DayView.PropertyChanged += OvViewPropertyChanged;
            scheduler.WorkWeekView.PropertyChanged += OvViewPropertyChanged;
            scheduler.FullWeekView.PropertyChanged += OvViewPropertyChanged;
            UpdateTrackBar();
        }

        public static void Detach(SchedulerControl scheduler) {
            scheduler.ActiveViewChanged -= OnSchedulerActiveViewChanged;
            scheduler.DayView.PropertyChanged -= OvViewPropertyChanged;
            scheduler.WorkWeekView.PropertyChanged -= OvViewPropertyChanged;
            scheduler.FullWeekView.PropertyChanged -= OvViewPropertyChanged;
            Scheduler = null;
            UpdateTrackBar();
        }

        static void OvViewPropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e) {
            if (e.PropertyName == "TimeScale")
                UpdateTrackBar();
        }

        static void OnTrackBarItemEditValueChanged(object sender, EventArgs e) {
            UpdateControl();
        }


        static void OnSchedulerActiveViewChanged(object sender, EventArgs e) {
            UpdateTrackBar();
        }

        static void UpdateControl() {
            if (Scheduler == null)
                return;
            if (TrackBarItem.EditValue == null)
                return;
            DayView dayView = Scheduler.ActiveView as DayView;
            if (dayView != null)
                dayView.TimeScale = dayView.TimeSlots[(int)TrackBarItem.EditValue].Value;
            TimelineView timelineView = Scheduler.ActiveView as TimelineView;
            if (timelineView != null)
                timelineView.GetBaseTimeScale().Width = (int)TrackBarItem.EditValue;
        }

        static void UpdateTrackBar() {
            if (TrackBarItem == null || TrackBarRepositoryItem == null)
                return;
            if (Scheduler == null) {
                TrackBarItem.Enabled = false;
                TrackBarRepositoryItem.Maximum = 0;
                TrackBarItem.EditValue = null;
                return;
            }
            DayView dayView = Scheduler.ActiveView as DayView;
            if (dayView != null) {
                TrackBarRepositoryItem.Minimum = 0;
                TrackBarRepositoryItem.Middle = 1;
                TrackBarRepositoryItem.Maximum = dayView.TimeSlots.Count - 1;
                TrackBarItem.EditValue = GetTimeSlotIndex(dayView.TimeSlots, dayView.TimeScale);
            }
            TimelineView timelineView = Scheduler.ActiveView as TimelineView;
            if (timelineView != null) {
                TrackBarRepositoryItem.Maximum = 100;
                TrackBarRepositoryItem.Minimum = 40;
                TrackBarItem.EditValue = timelineView.GetBaseTimeScale().Width;
            }
            TrackBarItem.Enabled = timelineView != null || dayView != null;
        }
        static int GetTimeSlotIndex(TimeSlotCollection timeSlots, TimeSpan scale) {
            for (int i = 0; i < timeSlots.Count; i++)
                if (timeSlots[i].Value == scale)
                    return i;
            return -1;
        }
    }
}
